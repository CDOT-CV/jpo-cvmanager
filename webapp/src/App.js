import React, { useEffect } from "react";
import { css } from "@emotion/react";
import RingLoader from "react-spinners/RingLoader";
import { GoogleOAuthProvider } from "@react-oauth/google";
import Header from "./components/Header";
import Menu from "./features/menu/Menu";
import Help from "./components/Help";
import Admin from "./pages/Admin";
import Grid from "@material-ui/core/Grid";
import Tabs from "./components/Tabs";
import Map from "./pages/Map";
import HeatMap from "./pages/HeatMap";
import WzdxMap from "./pages/WzdxMap";
import RsuMapView from "./pages/RsuMapView";
import BsmMap from "./pages/BsmMap";
import EnvironmentVars from "./EnvironmentVars";
import "./App.css";
import { useSelector, useDispatch } from "react-redux";
import {
  selectDisplayMap,

  // Actions
  getRsuData,
} from "./generalSlices/rsuSlice";
import { selectAuthLoginData, selectRole, selectLoadingGlobal } from "./generalSlices/userSlice";

const App = () => {
  const dispatch = useDispatch();

  const displayMap = useSelector(selectDisplayMap);
  const authLoginData = useSelector(selectAuthLoginData);
  const userRole = useSelector(selectRole);
  const loadingGlobal = useSelector(selectLoadingGlobal);

  useEffect(() => {
    // Refresh Data
    dispatch(getRsuData());
  }, [authLoginData, dispatch]);

  return (
    <GoogleOAuthProvider clientId={EnvironmentVars.GOOGLE_CLIENT_ID}>
      <div id="masterdiv">
        <Grid container id="content-grid" alignItems="center">
          <Header />
          {authLoginData ? (
            <Tabs>
              <div label="RSU Map">
                {displayMap ? null : <Menu />}
                {displayMap ? <RsuMapView auth={true} /> : <Map auth={true} />}
              </div>
              <div label="Heat Map">
                {displayMap ? null : <Menu />}
                <HeatMap auth={true} />
              </div>
              <div label="WZDx">
                <WzdxMap auth={true} />
              </div>
              <div label="BSM Map">
                <BsmMap auth={true} />
              </div>
              {userRole === "admin" && (
                <div label="Admin">
                  <Admin />
                </div>
              )}
              <div label="Help">
                <Help />
              </div>
            </Tabs>
          ) : (
            <div></div>
          )}
        </Grid>
        <RingLoader css={loadercss} size={200} color={"#13d48d"} loading={loadingGlobal} speedMultiplier={1} />
      </div>
    </GoogleOAuthProvider>
  );
};

const loadercss = css`
  display: block;
  margin: 0 auto;
  position: absolute;
  top: 50%;
  left: 50%;
  margin-top: -125px;
  margin-left: -125px;
`;

export default App;
