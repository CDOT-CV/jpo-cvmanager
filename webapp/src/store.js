import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
