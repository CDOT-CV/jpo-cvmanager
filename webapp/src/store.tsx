import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import intersectionMapReducer from './components/map/map-slice'
import intersectionMapLayerStyleReducer from './components/map/map-layer-style-slice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
    },
    preloadedState,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
