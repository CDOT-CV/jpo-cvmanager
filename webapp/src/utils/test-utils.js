/**
 * This function searches for the every react-aria SSR ids in a given HTMLElement node and replace every attribute values with a static id
 *
 * This can be usefull when you're trying to generate a snapshot of components using react-aria under the hood
 *
 * @ex :
 * ```
 * const { container } = render(<Component />);
 *
 * replaceReactAriaIds(container);
 * ```
 *
 * @param container The HTMLElement node to search for SSR ids
 */
function replaceChaoticIds(container) {
  const props = [
    {
      selector: "class",
      updateFunc: (val) =>
        val.replace(
          /css-[0-9a-z]{6}-MuiTableCell-root-MuiTablePagination-root/g,
          "css-mocked-MuiTableCell-root-MuiTablePagination-root"
        ),
    },
  ];

  container.querySelectorAll("td").forEach((item) => {
    props.forEach((prop) => {
      if (item.getAttribute(prop.selector)) {
        item.setAttribute(prop.selector, prop.updateFunc(item.getAttribute(prop.selector)));
      }
    });
  });
  container.querySelectorAll("input[aria-invalid]").forEach((item) => {
    delete item["aria-invalid"];
  });
  return container;
}

export { replaceChaoticIds };
