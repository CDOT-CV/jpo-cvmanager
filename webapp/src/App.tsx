import React, { useEffect } from 'react'
import { css } from '@emotion/react'
import RingLoader from 'react-spinners/RingLoader'
import Header from './components/Header'
import Menu from './features/menu/Menu'
import Help from './components/Help'
import Admin from './pages/Admin'
import Grid from '@material-ui/core/Grid'
import Tabs from './components/Tabs'
import Map from './pages/Map'
import RsuMapView from './pages/RsuMapView'
import './App.css'
import { useSelector, useDispatch } from 'react-redux'
import {
  selectDisplayMap,

  // Actions
  getRsuData,
  getRsuInfoOnly,
} from './generalSlices/rsuSlice'
import { selectAuthLoginData, selectRole, selectLoadingGlobal, setKcFailure } from './generalSlices/userSlice'
import { ReactKeycloakProvider } from '@react-keycloak/web'
import keycloak from './keycloak-config'
import { keycloakLogin } from './generalSlices/userSlice'
import { ThunkDispatch } from 'redux-thunk'
import { RootState } from './store'
import { AnyAction } from '@reduxjs/toolkit'

let loginDispatched = false

const App = () => {
  const dispatch: ThunkDispatch<RootState, void, AnyAction> = useDispatch()
  const displayMap = useSelector(selectDisplayMap)
  const authLoginData = useSelector(selectAuthLoginData)
  const userRole = useSelector(selectRole)
  const loadingGlobal = useSelector(selectLoadingGlobal)

  useEffect(() => {
    keycloak
      .updateToken(300)
      .then(function (refreshed) {
        if (refreshed) {
          console.debug('Token was successfully refreshed')
        } else {
          console.debug('Token is still valid')
        }
      })
      .catch(function () {
        // dispatch(setKcFailure(true))
        console.error('Failed to refresh the token, or the session has expired')
      })
  }, [])

  useEffect(() => {
    // Refresh Data
    console.debug('Authorizing the user with the API')
    dispatch(getRsuData({ test: 'test' }))
  }, [authLoginData, dispatch])

  return (
    <ReactKeycloakProvider
      initOptions={{ onLoad: 'login-required' }}
      authClient={keycloak}
      onTokens={({ token }) => {
        // Logic to prevent multiple login triggers
        if (!loginDispatched && token) {
          console.debug('onTokens loginDispatched:')
          dispatch(keycloakLogin(token))
          loginDispatched = true
        }
        setTimeout(() => (loginDispatched = false), 5000)
      }}
    >
      <div id="masterdiv">
        <Grid container id="content-grid" alignItems="center">
          <Header />
          {authLoginData && keycloak?.authenticated ? (
            <Tabs>
              <div label="RSU Map">
                {displayMap ? null : <Menu />}
                {displayMap ? <RsuMapView auth={true} /> : <Map auth={true} />}
              </div>
              {userRole === 'admin' && (
                <div label="Admin">
                  <div label="Admin">
                    <Admin updateRsuData={() => dispatch(getRsuInfoOnly())} />
                  </div>
                </div>
              )}
              <div label="Help">
                <Help />
              </div>
            </Tabs>
          ) : (
            <div></div>
          )}
        </Grid>
        <RingLoader css={loadercss} size={200} color={'#13d48d'} loading={loadingGlobal} speedMultiplier={1} />
      </div>
    </ReactKeycloakProvider>
  )
}

const loadercss = css`
  display: block;
  margin: 0 auto;
  position: absolute;
  top: 50%;
  left: 50%;
  margin-top: -125px;
  margin-left: -125px;
`

export default App
