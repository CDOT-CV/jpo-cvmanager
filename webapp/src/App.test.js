import React from 'react'
import { render } from '@testing-library/react'
import App from './App'
import { Provider } from 'react-redux'
import { setupStore } from './store'
import { replaceChaoticIds } from './utils/test-utils'

it('should take a snapshot', () => {
  const { container } = render(
    <Provider
      store={setupStore({
        user: {
          value: {
            authLoginData: { data: 'data' },
          },
        },
      })}
    >
      <App />
    </Provider>
  )

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})
