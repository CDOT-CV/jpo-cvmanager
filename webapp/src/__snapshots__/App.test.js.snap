// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div
    id="masterdiv"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
      id="content-grid"
    >
      <div>
        <header
          id="header"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
          >
            <img
              alt="Logo"
              id="logo"
              src="cdot_logo.png"
            />
            <h1
              id="header-text"
            >
              CDOT CV Manager
            </h1>
            <div
              id="login"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
              >
                <div
                  class="MuiGrid-root"
                  id="userInfoGrid"
                >
                  <h3
                    id="nameText"
                  />
                  <h3
                    id="emailText"
                  />
                  <select
                    id="organizationDropdown"
                  />
                </div>
                <button
                  id="logout"
                >
                  Logout
                </button>
              </div>
            </div>
          </div>
        </header>
      </div>
      <div
        class="tabs"
      >
        <ol
          class="tab-list"
        >
          <li
            class="tab-list-item tab-list-active"
          >
            RSU Map
          </li>
          <li
            class="tab-list-item"
          >
            Help
          </li>
        </ol>
        <div
          class="tab-content"
        >
          <div>
            <div>
              <button
                id="toggle"
              >
                Display Counts
              </button>
            </div>
          </div>
          <div
            class="container"
          >
            <div
              class="MuiGrid-root legend-grid MuiGrid-container MuiGrid-align-items-xs-center"
            >
              <div
                class="legend"
              >
                <h1
                  class="legend-header"
                >
                  Map Layers
                </h1>
                <div
                  class="legend-item"
                >
                  <label
                    class="legend-label"
                  >
                    <input
                      checked=""
                      class="legend-input"
                      type="checkbox"
                    />
                    RSU
                  </label>
                </div>
                <div
                  class="legend-item"
                >
                  <label
                    class="legend-label"
                  >
                    <input
                      class="legend-input"
                      type="checkbox"
                    />
                    Heatmap
                  </label>
                </div>
                <div
                  class="legend-item"
                >
                  <label
                    class="legend-label"
                  >
                    <input
                      class="legend-input"
                      type="checkbox"
                    />
                    BSM Viewer
                  </label>
                </div>
                <div
                  class="legend-item"
                >
                  <label
                    class="legend-label"
                  >
                    <input
                      class="legend-input"
                      type="checkbox"
                    />
                    WZDx
                  </label>
                </div>
              </div>
              <div
                class="rsu-status-div"
              >
                <h1
                  class="legend-header"
                >
                  RSU Status
                </h1>
                <label
                  class="rsu-status-label"
                >
                  <input
                    checked=""
                    class="rsu-status-input"
                    name="none-status-radio"
                    type="radio"
                    value="none"
                  />
                  None
                </label>
                <label
                  class="rsu-status-label"
                >
                  <input
                    class="rsu-status-input"
                    name="online-status-radio"
                    type="radio"
                    value="online"
                  />
                  Online Status
                </label>
                <label
                  class="rsu-status-label"
                >
                  <input
                    class="rsu-status-input"
                    name="scms-status-radio"
                    type="radio"
                    value="scms"
                  />
                  SCMS Status
                </label>
              </div>
            </div>
            <div
              class="container-fluid"
              style="width: 100%; height: calc(100vh - 136px); display: flex;"
            >
              <div
                style="position: relative; width: 100%; height: 100%;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <span
      class="css-gv9u3i"
    >
      <span
        class="css-1t15js"
      />
      <span
        class="css-15ktaxl"
      />
    </span>
  </div>
</div>
`;
